### PR TITLE
Add Python Client to Test Negative HTTP2 Conditions

### PR DESCRIPTION
**Pull request by ncteisen**
These tests are for the new Http/2 test server (#8900). They are designed to test client behavior when the server sends unexpected responses during rpcs.

cc @ericgribkoff @makdharma

### CODE REVIEW

**Comment by nathanielmanistaatgoogle** [comment]
Cool! Please prepend a leading underscore on the file name. Does this test need to be recorded in `tests.json`?

**Comment by ncteisen** [comment]
Renamed the file. Although this file is more similar to interop/client.py than is is to the unit tests, I was wondering if maybe it should follow that convention instead?

And at this point, no, it does not belong in tests.json since it cannot run as a standalone test. Eventually, we will do the same thing to is as was done to the interop tests, by making them into unit tests.

**Comment by nathanielmanistaatgoogle** [comment]
You're absolutely right - the file should not have a leading underscore and should not be added to `tests.json`; I apologize for the erroneous comment.

Please link in your commit log message to the document specifying these tests? I expect it would be a sibling of [the interop test specifications](https://github.com/grpc/grpc/blob/master/doc/interop-test-descriptions.md)?

**Comment by ncteisen** [comment]
The documentation is in progress. Will link to it and update this thread when it is finished.

File: src/python/grpcio_tests/tests/http2/_negative_http2_client.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    raise ValueError(
+      'expected message %s, got %s' % (expected_details, call.details()))
+
+def _validate_status_code_and_details(call, expected_code, expected_details):
+  _expect_status_code(call, expected_code)
+  _expect_status_details(call, expected_details)
+
+# common requests
+_SMALL_SIZE = 1024
+_LARGE_SIZE = 314159
```
Comment (adelez) [inline src/python/grpcio_tests/tests/http2/_negative_http2_client.py]: This is not used anywhere.

File: src/python/grpcio_tests/tests/http2/_negative_http2_client.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+_SMALL_SIMPLE_REQUEST = messages_pb2.SimpleRequest(
```
Comment (adelez) [inline src/python/grpcio_tests/tests/http2/_negative_http2_client.py]: According to the doc https://github.com/ncteisen/grpc/blob/0eb69186f0b22f0e4bd5d6f1b4e31252bf614b76/doc/negative-http2-interop-test-descriptions.md, the request size is large...

File: src/python/grpcio_tests/tests/http2/_negative_http2_client.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    response_type=messages_pb2.COMPRESSABLE,
+    response_size=_SMALL_SIZE,
```
Comment (adelez) [inline src/python/grpcio_tests/tests/http2/_negative_http2_client.py]: According to the same doc the response_size should be _LARGE_SIZE. Can you please review the sizes to make sure they are in sync with the doc. Thanks!

File: src/python/grpcio_tests/tests/http2/_negative_http2_client.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+  for _ in range(15):
+    futures.append(stub.UnaryCall.future(_SMALL_SIMPLE_REQUEST))
+  for future in futures:
+    _validate_payload_type_and_length(future.result(),
+        messages_pb2.COMPRESSABLE, _SMALL_SIZE)
+
+def _run_test_case(test_case, stub):
+  if test_case == 'goaway':
+    _goaway(stub)
+  elif test_case == 'rst_after_header':
```
Comment (adelez) [inline src/python/grpcio_tests/tests/http2/_negative_http2_client.py]: Java client uses all upper case letters for the test cases (https://github.com/grpc/grpc-java/pull/2513/files). Let's be consistent.

**Comment by ncteisen** [comment]
Addressed all comments. Documentation is [here](https://github.com/grpc/grpc/blob/master/doc/negative-http2-interop-test-descriptions.md).

**Comment by ncteisen** [comment]
Almost all test are green. cloud_to_cloud:ruby:node_server:timeout_on_sleeping_server fails, but that is a well known flake (see #8989).

Going to merge
